modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequestBuilder.java

```diff
@@ -0,0 +1,121 @@
+import org.elasticsearch.action.ActionRequestBuilder;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.WriteConsistencyLevel;
+import org.elasticsearch.action.search.SearchRequestBuilder;
+import org.elasticsearch.client.ElasticsearchClient;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.query.QueryBuilder;
+
+public abstract class AbstractBulkByScrollRequestBuilder<
+                Request extends AbstractBulkByScrollRequest<Request>,
+                Response extends ActionResponse,
+                Self extends AbstractBulkByScrollRequestBuilder<Request, Response, Self>>
+        extends ActionRequestBuilder<Request, Response, Self> {
+    private final SearchRequestBuilder source;
+
+    protected AbstractBulkByScrollRequestBuilder(ElasticsearchClient client,
+            Action<Request, Response, Self> action, SearchRequestBuilder source, Request request) {
+        super(client, action, request);
+        this.source = source;
+    }
+
+    protected abstract Self self();
+
+    /**
+     * The search used to find documents to process.
+     */
+    public SearchRequestBuilder source() {
+        return source;
+    }
+
+    /**
+     * Set the source indices.
+     */
+    public Self source(String... indices) {
+        source.setIndices(indices);
+        return self();
+    }
+
+    /**
+     * Set the query that will filter the source. Just a convenience method for
+     * easy chaining.
+     */
+    public Self filter(QueryBuilder filter) {
+        source.setQuery(filter);
+        return self();
+    }
+
+    /**
+     * The maximum number of documents to attempt.
+     */
+    public Self size(int size) {
+        request.setSize(size);
+        return self();
+    }
+
+    /**
+     * Should we version conflicts cause the action to abort?
+     */
+    public Self abortOnVersionConflict(boolean abortOnVersionConflict) {
+        request.setAbortOnVersionConflict(abortOnVersionConflict);
+        return self();
+    }
+
+    /**
+     * Call refresh on the indexes we've written to after the request ends?
+     */
+    public Self refresh(boolean refresh) {
+        request.setRefresh(refresh);
+        return self();
+    }
+
+    /**
+     * Timeout to wait for the shards on to be available for each bulk request.
+     */
+    public Self timeout(TimeValue timeout) {
+        request.setTimeout(timeout);
+        return self();
+    }
+
+    /**
+     * Consistency level for write requests.
+     */
+    public Self consistency(WriteConsistencyLevel consistency) {
+        request.setConsistency(consistency);
+        return self();
+    }
+
+    @Override
+    public Request request() {
+        return beforeExecute(request);
+    }
+    
+    @Override
+    protected Request beforeExecute(Request request) {
+        // In 2.x the search's "source" isn't built until 
+        request.getSearchRequest().source(source.internalBuilder());
```

Comment:
Just like all the request builders that interact with search requests in 2.x, this builder must build the source explicitly or the request is totally empty and everything fails. This is done both here and above, in the `request` method, because that is how SearchRequestBuilder does it. It works though I'm not totally sure why we have to places. I don't lose sleep over it because it is a 2.x hack.